src/nvim/mark_extended.c

```diff
@@ -0,0 +1,318 @@
+ExtendedMark *extmark_prev(buf_T *buf, char *ns, pos_T *pos)
+{
+  return extmark_neighbour(buf, ns, pos, 0);
+}
+
+ExtmarkArray *extmark_nextrange(buf_T *buf, char *ns, pos_T *lower, pos_T *upper)
+{
+  return extmark_neighbour_range(buf, ns, lower, upper, 1);
+}
+
+/* Returns the position of marks between a range, */
+/* marks found at the start or end index will be included, */
+/* if upper.lnum or upper.col are negative the buffer */
+/* will be searched to the end, */
+/* go_forward can be set to control the order of the array */
+static ExtmarkArray *extmark_neighbour_range(buf_T *buf, char *ns, pos_T *lower, pos_T *upper, bool go_forward)
+{
+  ExtmarkArray *array = (ExtmarkArray *)xmalloc(sizeof(ExtmarkArray));
+  kv_init(*array);
+  int cmp;
+
+  FOR_EXTMARKS_IN_NS(buf, ns)
+    if (pos_cmp(*lower, *upper) == 1) {
+      break;
+    }
+    cmp = pos_cmp(*lower, extmark->fmark.mark);
+    if (cmp != 1) {
+      kv_push(*array, extmark);
+    }
+    else if (cmp == 1) {
+      break;
+    }
+  END_FOR_EXTMARKS_IN_NS
+
+  /* Swap the elements in the array */
+  /* if (!go_forward) { */
+      /* ExtendedMark holder; */
+      /* int n = (int)sizeof(array); */
+      /* int end = n-1; */
+      /* for (int i = 0; i < n/2; i++) { */
+          /* holder = kv_A(*array, i); */
+          /* array[i] = array[end]; */
+          /* array[end] = holder; */
+          /* end --; */
+      /* } */
+  /* } */
+  return array;
+}
+
+/* Returns the mark, nearest to the passed in mark*/
+static ExtendedMark *extmark_neighbour(buf_T *buf, char *ns, pos_T *input, bool go_forward)
+{
+  int cmp;
+  if (go_forward) {
+    FOR_ALL_EXTMARKS(buf)
+      cmp = pos_cmp(extmark->fmark.mark, *input);
+      if (cmp == 1) {
+        return extmark;
+      }
+    END_FOR_ALL_EXTMARKS
+  } else {
+    FOR_ALL_EXTMARKS_WITH_PREV(buf)
+      cmp = pos_cmp(extmark->fmark.mark, *input);
+      if (cmp == 1) {
+        return prev;
+      }
+    END_FOR_ALL_EXTMARKS_WITH_PREV
+  }
+  return NULL;
+}
+int called=0;
+static bool extmark_create(buf_T *buf, char* ns, char *name, linenr_T row, colnr_T col)
+{
+  /* Initialize as first mark on this buffer*/
+  called ++;
+  if (!buf->b_extmarks) {
+    buf->b_extmarks = pmap_new(cstr_t)();
+    buf->b_extmarks_tree = kb_init(extmarks, KB_DEFAULT_SIZE);
+  }
+  StringMap *ns_map = pmap_get(cstr_t)(buf->b_extmarks, (cstr_t)ns);
+  /* Initialize a new namespace */
+  /* if (!ns_map || !kh_size(ns_map->table)) { */
+  ns = xstrdup(ns);
+  if (!ns_map) {
+    ns_map = pmap_new(cstr_t)();
+    pmap_put(cstr_t)(buf->b_extmarks, (cstr_t)ns, ns_map);
+  }
+
+  name = xstrdup(name);
+  ExtendedMark *extmark = (ExtendedMark *)xmalloc(sizeof(ExtendedMark));
+  /* extmark->fmark.mark = gen_relative(buf, ns, row, col); */
+  extmark->fmark.mark.lnum = row;
+  extmark->fmark.mark.col = col;
+  extmark->names = pmap_new(cstr_t)();
+  pmap_put(cstr_t)(extmark->names, (cstr_t)ns, &name);
+
+  kb_put(extmarks, buf->b_extmarks_tree,  *extmark);
```

Comment:
Here is the thing: a pointer is also a value. The kb_tree api is not very convenient for this use case, but it works. Here is a sketch: (I have not tested it)

```
typedef struct {
    pos_T key;
    ... more data ..
} elem_t;

#define elem_cmp(a, b) (pos_cmp(a->key, b->key))
KBTREE_INIT(test, elem_t, elem_cmp)

static elem_t *insert_elem(bktree_t(test) *b, pos_T pos, bool* is_new) {
    elem_t t, *p , **pp;
    t.key = pos;
    pp = kb_get(test, b, &t);
    // IMPORTANT: put() only works if key is absent
    if (pp) {
        *is_new = false;
        return *pp;
    }
    p = xmalloc(sizeof(*p));
    p->key = pos;
    kb_put(test, b, p);
    *is_new = true;
    return p;
}
```